generator/internal/stub_generator.cc

```diff
@@ -0,0 +1,134 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "generator/internal/stub_generator.h"
+#include "absl/memory/memory.h"
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_replace.h"
+#include "absl/strings/str_split.h"
+#include "absl/strings/strip.h"
+#include "generator/internal/codegen_utils.h"
+#include "generator/internal/printer.h"
+#include <google/api/client.pb.h>
+#include <google/protobuf/descriptor.h>
+
+namespace google {
+namespace cloud {
+namespace generator_internal {
+
+StubGenerator::StubGenerator(
+    google::protobuf::ServiceDescriptor const* service_descriptor,
+    std::map<std::string, std::string> service_vars,
+    google::protobuf::compiler::GeneratorContext* context)
+    : service_descriptor_(service_descriptor),
+      vars_(std::move(service_vars)),
+      header_(context, vars_["stub_header_path"]),
+      cc_(context, vars_["stub_cc_path"]) {
+  SetVars();
+}
+
+Status StubGenerator::GenerateHeader() {
+  // Temporary conditional for unit testing purposes.
+  if (service_descriptor_->name() == "FailureService") {
+    return Status(StatusCode::kInternal, "Failed for testing.");
+  }
+
+  header_.Print(vars_,  // clang-format off
+    "// Generated by the Codegen C++ plugin.\n"
```

Comment:
Do generated files need to have copyright comments at the top? I don't know, but I'm wondering. What do grpc's or proto-generated files look like? Those might be different because those aren't checked in.

... Here's one example from checked-in golang code, that does include a copyright header: https://github.com/googleapis/google-cloud-go/blob/master/container/apiv1/cluster_manager_client.go

Maybe we should do that too.